core/src/main/java/de/danoeh/antennapod/core/service/download/DownloadService.java

```diff
@@ -550,14 +549,9 @@ private void stopServiceIfEverythingDone() {
 
         if (numberOfDownloads.get() <= 0 && DownloadRequester.getInstance().hasNoDownloads()) {
             Log.d(TAG, "Attempting shutdown");
+            cancelNotificationUpdater();
             stopForeground(true);
             stopSelf();
-
-            // Trick to hide the notification more quickly when the service is stopped
-            // Without this, the second-last update of the notification stays for 3 seconds after onDestroy returns
-            notificationUpdater.run();
-            NotificationManager nm = (NotificationManager) getSystemService(NOTIFICATION_SERVICE);
-            nm.cancel(R.id.notification_downloading);
```

Comment:
Thanks for looking into this! Actually, these lines were added deliberately to hide the notification more quickly. I know it's a bit strange but when *showing* the notification (even without the `cancel` call), the Android system *hides* the notification more quickly. Is moving `cancelNotificationUpdater` not enough to fix the problem?

Reply:
No. I added logs and cancelNotificationUpdate acted as a race condition (only happening sometimes).

This "show" just before the cancel reliably triggers the issue though. Certainly on the emulator (see my comments on the issue). I can test on my phone in a real scenario but it'll take some days due to how I use AntennaPod.

I didn't notice any slowness in hiding of the notification. Maybe you could use stopForeground(false) and then cancel the notification? But that's for lollipop and up only.

Reply:
> I didn't notice any slowness in hiding of the notification.

See this (crappy - sorry) screen recording on an API30 emulator. 
https://drive.google.com/file/d/1Jv1tuYzcmJwH0c7_3NLW7eNJQpRp5h3_/view?usp=sharing

The notification stays there for some seconds even after the service already went through `onDestroy` (verified using logcat). It also displays the podcast that already finished loading, even though the notification updater was already called to change the text to "Processing downloads". With the workaround, the notification is hidden directly. Not sure why this happens... :/ Before introducing the workaround, we got comments by users who were confused by the fact that notification and progress bar disagree about whether it is finished or not.

Reply:
Ah... I didn't compare the spinner to the notification :)

I have an api 30 device and emulator so I'll see if I can make it behave.

Reply:
I do not see a several-second delay, but I do see a delay. The delay I'm seeing though comes as a side effect of scheduling the UI update (by calling removeDownload(downloader)) separately to calling stopForeground.

I have updated the code to call cancelNotificationUpdater and stopForeground before the other events. It seems that closing a notification is a slow, asynchronous process so the UI still manages to update first, but the delay is reduced for me. I think that this should be safe as onStartCommand has a call to startForeground and setupNotificationUpdaterIfNecessary so in the event that the notification is closed just as a new command is being sent, a new notification and updater thread will be created.

Reply:
ooh... I just saw the long delay on my phone. How odd...

Reply:
This appears to be why the notification doesn't go away?!

```
2021-02-27 17:13:39.560 13176-13176/de.danoeh.antennapod.debug I/System.out: link - start a worker with a 5 second delay
2021-02-27 17:13:39.564 13176-13176/de.danoeh.antennapod.debug D/DBTasks: autodownloadUndownloadedItems
2021-02-27 17:13:39.564 13176-13176/de.danoeh.antennapod.debug I/System.out: link - end of onDestroy
2021-02-27 17:13:39.565 13176-13276/de.danoeh.antennapod.debug D/NetworkUtils: Device is connected to Wi-Fi
2021-02-27 17:13:39.565 13176-13276/de.danoeh.antennapod.debug D/NetworkUtils: Auto-dl filter is disabled
2021-02-27 17:13:44.623 13176-13277/de.danoeh.antennapod.debug I/WM-WorkerWrapper: Worker result SUCCESS for Work [ id=a80d19f3-93c8-43d5-98a6-ef2ec1018c1d, tags={ de.danoeh.antennapod.core.sync.SyncService } ]
```

When that last line prints is when I see the notification disappear. And that is true even with the "fix" of showing the notification and clearing it again.

This applies on my phone if I refresh a single feed when there are no new episodes. I do not see this in the emulator.

Reply:
I removed the 5 seconds on that worker and the notification still gets stuck, so ... I guess it's not related to that :(

Reply:
Yeah, I think the worker should be unrelated. It's only responsible for gpodder sync.

Reply:
Ugh. I have trying various permutations of code and I think I have narrowed it down to these mutually-exclusive cases:
1. If you remove the workaround, the notification may close slowly. (This was primarily visible to me when refreshing a podcast that had no changes - subscribing to a regular podcast is how I tested this).
2. If you add the workaround, the notification may get stuck. (Yesterday, this was primarily visible to me when starting the app cold, refreshing all podcasts and having new episodes show up - the yasmar.net podcasts are how I tested this).

This holds both on my phone (Pixel 4a running Android 11) and emulator (API 30).

I'm not quite sure why, but the second issue was not being reliable today. When I did trigger the second issue, I found that no amount of delayed closing seemed to help. It's almost like the system thinks the notification is not there, even though it is. I wonder if both issues are caused by the same bug, but since in case 1 the notification is tied to the service, the system manages to clean it up eventually.

I did observe that low-level logs (from NotificationService and NotificationHistory) were spit out when the workaround was in place (ie. when the notification was shown not tied to the service), but these logs did not print when issue 2 happened. I'm not sure what to make of that.

```
2021-02-27 20:57:11.402 521-521/system_process D/NotificationService: 0|de.danoeh.antennapod.debug|2131362257|null|10156: granting content://settings/system/notification_sound
2021-02-27 20:57:11.403 521-521/system_process D/NotificationService: 0|de.danoeh.antennapod.debug|2131362257|null|10156: granting content://settings/system/notification_sound
2021-02-27 20:57:11.417 521-521/system_process W/NotificationHistory: Attempted to add notif for locked/gone/disabled user 0
```

Of these 2 issues, I think that number 2 is worse, since it leaves behind a permanent problem that requires active steps to fix while the first issue merely results in several more seconds of a notification being visible.


Since I have tried and failed a front-on assault... maybe a different approach is required here?

eg.
1. Start DownloadService as a background service.
2. Promote the running DownloadService to a foreground service if the app closes.
3. Demote the running DownloadService to a background service if the app opens.

Possibly in the scheduled fetch case, the service may need to start as a foreground service (to avoid the launch being blocked). In that case it might be simpler to always start it as a foreground service but immediately demote it if the app was visible when it launched.

Reply:
I know what the problem is, I think.

It looks to me that Foreground services cannot have a notification that is visible for very brief periods of time. That's why the "slow notification closing" is only visible for me when I refresh a single podcast that hasn't changed. I suspect this is so that the user can actually see sneaky apps running their services as foreground services to do things in the background. I have actually noticed this myself for some apps (that previously had very quick notification show/hide and changed to a longer show period) but I think it started in Android 10.

The thing is, this applies even if you start as foreground and immediately drop that. So for DownloadService initiated from the UI, it must start and remain a background service in order to be able to close the notification quickly. I tested this (by removing the foreground stuff completely) and it worked but that probably breaks the timed fetching.

While the workaround of "show another notification and then close that" that you were using works now, I wonder if Google might block that, since it seems to avoid what looks like a policy decision. And of course, there's that pesky bug that leaves the notification stuck. Maybe inserting delays can help here but that would seem to defeat the purpose of "hide as quickly as possible" :)

Reply:
> It looks to me that Foreground services cannot have a notification that is visible for very brief periods of time. That's why the "slow notification closing" is only visible for me when I refresh a single podcast that hasn't changed.

Good observation! How about changing the `Processing downloads` text to something like `Stopping service...`? Then it is clear what is happening. Then, just before calling `stopForeground`, call `notificationUpdater.run` a last time.

> Demote the running DownloadService to a background service if the app opens.

To be honest, I wouldn't do that. I would prefer to keep the changes as small as possible for now, so they do not introduce new issues. Then we can release the beta soon. The download service needs a rework anyway for Android 12 compatibility (see #4958), so thinking of all edge-cases here is basically wasted time. (by the way, in early versions of AntennaPod, the playback service also had a foreground/background switching logic. It caused quite some crashes)

Reply:
So... I think I should have been pushing branches with the ... branching ideas, so I've gone ahead and done that.

The basic bugfix for the stuck notification is in https://github.com/a1291762/AntennaPod/tree/bugfix/stuck-notification (https://github.com/a1291762/AntennaPod/commit/fcbbaa40603a04ba809058327c8c9620f3798570). I have duplicated this on develop (since that's where this PR comes from).

Changing the text ... that's going to require all the translations to change too, so this is an incomplete change: https://github.com/a1291762/AntennaPod/tree/idea/change-finished-text (https://github.com/a1291762/AntennaPod/commit/1c7c9f18bff0f31fb4d1067f0e1b62220dfee2fd). I guess the translation could change without the string id but that feels a bit wrong?

I actually did the background service thing yesterday and it was much simpler than I expected. But if you don't want it, that's fine too: https://github.com/a1291762/AntennaPod/tree/idea/background-download-service (https://github.com/a1291762/AntennaPod/commit/e1f13ed33e3c172d1e2ccd7fafe723a6d46caa24).

Reply:
> that's going to require all the translations to change too

That's not a problem - will be handled by the translators :) Even if it is not translated before the release, I think it is still easier to understand than the old text. So it would be great if you could push the commit to this PR :)

> I actually did the background service thing yesterday and it was much simpler than I expected.

I agree, it looks simpler than I expected, too. I would prefer to be better safe than sorry in this case and live with the longer notification for now.